x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/RealmConfigBootstrapCheck.java

```diff
@@ -0,0 +1,143 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.xpack.security;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.bootstrap.BootstrapCheck;
+import org.elasticsearch.bootstrap.BootstrapContext;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.threadpool.ThreadPool;
+import org.elasticsearch.watcher.ResourceWatcherService;
+import org.elasticsearch.xpack.core.security.SecurityExtension;
+import org.elasticsearch.xpack.core.security.authc.Realm;
+import org.elasticsearch.xpack.core.security.authc.RealmConfig;
+import org.elasticsearch.xpack.core.security.authc.RealmSettings;
+import org.elasticsearch.xpack.core.security.authc.esnative.NativeRealmSettings;
+import org.elasticsearch.xpack.core.security.authc.file.FileRealmSettings;
+import org.elasticsearch.xpack.security.authc.InternalRealms;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Performs bootstrap checks required across security realms using realm
+ * configuration.<br>
+ * The checks also run for realms provided by security extensions.<br>
+ * Not all checks could be performed with bootstrap check, the kind of checks
+ * performed are like invalid/missing types, allowed no of instances for realm
+ * type or lookup realms validation.
+ */
+public class RealmConfigBootstrapCheck implements BootstrapCheck {
+    private static final Logger LOGGER = Loggers.getLogger(RealmConfigBootstrapCheck.class);
+
+    private final List<SecurityExtension> securityExtensions;
+
+    public RealmConfigBootstrapCheck(final List<SecurityExtension> securityExtensions) {
+        this.securityExtensions = Collections.unmodifiableList(securityExtensions);
+    }
+
+    @Override
+    public BootstrapCheckResult check(BootstrapContext context) {
+        BootstrapCheckResult result = BootstrapCheckResult.success();
+        final ThreadPool threadPool = new ThreadPool(context.settings);
+        try (ResourceWatcherService watcherService = new ResourceWatcherService(context.settings, threadPool)) {
+            final Map<String, Realm.Factory> extensionsRealmTypes = new HashMap<>();
+            securityExtensions.stream().forEach(e -> extensionsRealmTypes.putAll(e.getRealms(watcherService)));
```

Comment:
We don't want to do this - it means this method changes from being called _once_ to being called _twice_.

I think this is over complicating the check.
Just run through the list of realms from `RealmSettings.getRealmSettings()`, and then create three `Set`s
1. `delegatingRealms` : the names of realms that have a value for `lookup_realms`
2. `targetRealms` : the names of realms that occur in `lookup_realms`  and in `allRealms`
3. `missingRealms` : the names of realms that occur in `lookup_realms`  but not in `allRealms`

Then 
(a) find intersection of `delegatingRealms` and `targetRealms` sets (that is, realm names that are both the source of a lookup and the target of a lookup) If that intersection is non empty, then you have an error.
(b) If `missingRealms` is non-empty you have another error.